### PR TITLE
Clientinfo authentication

### PR DESCRIPTION
**Pull request by zmstone**
https://emqx.atlassian.net/browse/EMQX-12626

Release version: v/e5.8.1

## Summary

ClientInfo based authentication is a light-weight rule-based authentication mechanism.
It is useful in below cases.

- As a fencing mechanism for invalid/intrusive clients.
  For example, if valid clients should all start with a prefix, there should be no need to check other authenticators (which are usually heavier)
- As an extended validation for client misconfiguration.
  For example, after client certificate is validated (mTLS), it might be enough to check if the client ID is the same as certificate common name.
- A secret algorithm or a secret to validate password.
  For example, `password = sha256(clientid + secret)`

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [x] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket https://github.com/emqx/emqx-docs/pull/2670 and https://github.com/emqx/emqx-docs/pull/2671
- [x] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx_auth_cinfo/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Authenticate clients with connection information
+
+This application implements an extended authentication for EMQX Enterprise edition.
+
+Client-info (of type `cinfo`) authentication is a light-weight authentication mechanism which checks client properties and attributes against user defined rules.
```
Comment (qzhuyan) [inline apps/emqx_auth_cinfo/README.md:5]: nit, lightweight

File: apps/emqx_auth_cinfo/src/emqx_auth_cinfo_app.erl [hunk 2]
```diff
@@ -0,0 +1,20 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2020-2024 EMQ Technologies Co., Ltd. All Rights Reserved.
```
Comment (qzhuyan) [inline apps/emqx_auth_cinfo/src/emqx_auth_cinfo_app.erl:2]: nit, 2024

**Review by qzhuyan — DISMISSED** [review]

**Review by qzhuyan — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-4079)

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-4080)

File: rel/config/ee-examples/cinfo-authn.conf [hunk 3]
```diff
@@ -0,0 +1,30 @@
+authentication = [
+  {
+    mechanism = cinfo
+    checks = [
+      # allow clients with username starts with 'super-'
+      {
+        is_match = "regex_match(username, 'super-')"
```
Comment (Le9Wang) [inline rel/config/ee-examples/cinfo-authn.conf:7]: is_match = "regex_match(username, '^super-')"

File: apps/emqx_auth_cinfo/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  # ... more authentications ...
+  # ...
+  # if all authenticators are exhausted without an 'allow' or a 'deny' result, the client is not rejected
+]
+```
+
+More match expression examples:
+
+- TLS certificate common name is the same as username: `str_eq(cert_common_name, username)`
+- Password is the `sha1` hash of environment variable `EMQXVAR_SECRET` concatenated to client ID: `str_eq(password, hash(sha1, concat([clientid, getenv('SECRET')])))`
```
Comment (Le9Wang) [inline apps/emqx_auth_cinfo/README.md:39]: `str_eq(password, hash('sha1', concat([clientid, getenv('SECRET')])))`

Comment (zmstone) [reply apps/emqx_auth_cinfo/README.md:39]: thank you. fixed.

File: rel/config/ee-examples/cinfo-authn.conf (hunk 3 above)
Comment (zmstone) [reply rel/config/ee-examples/cinfo-authn.conf:7]: thank you, fixed.

**Pull request merged**